### PR TITLE
DinoCatalog v3

### PR DESCRIPTION
**Pull request by Manbearpixel**
Resubmitting code refactoring from yesterday along with two, hopefully fixed, odd Hound warnings.

### CODE REVIEW

File: dino_catalog/Dinosaur.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  def initialize(args)
+    args.each do |key, value|
+      value = normalize_data(value)
+      instance_variable_set("@#{key}", value)
+    end
+  end
+
+  def normalize_data(data = "")
+    return data if data.nil? || DATA_CONVERSIONS[(data.to_sym)].nil?
```
Comment (djkotowski) [inline dino_catalog/Dinosaur.rb]: Unnecessary parentheses around `data.to_sym`

File: dino_catalog/Dinosaur.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
+class Dinosaur
+  attr_accessor :name, :continent, :period, :weight, :walking, :description, :diet
+
+  DATA_CONVERSIONS = {
+    Yes:  "Carnivore",
+    No:   "Herbivore",
+  }
+
+  def initialize(args)
```
Comment (djkotowski) [inline dino_catalog/Dinosaur.rb]: Can you think of a better name than `args`?

File: dino_catalog/Dinosaur.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def larger_than?(size)
+    return false if weight.nil?
+    Float(weight) > Float(size)
+  end
+
+  def smaller_than?(size)
+    return false if weight.nil?
+    !larger_than?(size)
```
Comment (djkotowski) [inline dino_catalog/Dinosaur.rb]: This returns `true` if `weight == size`

File: dino_catalog/Dinosaur.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  def export_facts(facts = "")
+    to_hash.each do |header, value|
+      next if value.empty?
+      header = header.gsub(/[A-Za-z']+/, &:capitalize)
+      facts << sprintf("%-15s %s\n", header, value)
+    end
+    facts
+  end
+
+  def to_hash
```
Comment (djkotowski) [inline dino_catalog/Dinosaur.rb]: To keep consistent with the language, this should be `to_h`.  `to_hash` often is used for generating hash codes.

File: dino_catalog/dino_catalog.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+  def filter(key, value)
+    if key.to_s == "weight" && value.include?("<")
+      dinosaurs.select! { |dinosaur| dinosaur.smaller_than?(value.to_s[1, value.length]) }
+    elsif key.to_s == "weight" && value.include?(">")
+      dinosaurs.select! { |dinosaur| dinosaur.larger_than?(value.to_s[1, value.length]) }
+    else
+      dinosaurs.select! { |dinosaur| dinosaur.instance_variable_get(("@#{key}")).to_s.downcase.include?(value.downcase) }
+    end
+    self
```
Comment (djkotowski) [inline dino_catalog/dino_catalog.rb]: Nice way to support chaining.

File: dino_catalog/dino_catalog.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    @dinosaurs = dinosaurs.clone
+  end
+
+  def filter(key, value)
+    if key.to_s == "weight" && value.include?("<")
+      dinosaurs.select! { |dinosaur| dinosaur.smaller_than?(value.to_s[1, value.length]) }
+    elsif key.to_s == "weight" && value.include?(">")
+      dinosaurs.select! { |dinosaur| dinosaur.larger_than?(value.to_s[1, value.length]) }
+    else
+      dinosaurs.select! { |dinosaur| dinosaur.instance_variable_get(("@#{key}")).to_s.downcase.include?(value.downcase) }
```
Comment (djkotowski) [inline dino_catalog/dino_catalog.rb]: This comparison should be a method on your Dinosaur class.

File: dino_catalog/dino_catalog.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  def print_data(requested_data)
+    requested_data = Array(requested_data)
+    dinosaurs.each do |dinosaur|
+      dino_data ||= ''
+      requested_data.each do |header_name|
+        dino_data << "#{dinosaur.instance_variable_get(("@#{header_name}"))} | "
+      end
+      puts dino_data.slice(0..-4)
```
Comment (djkotowski) [inline dino_catalog/dino_catalog.rb]: This is a bit hard to follow.  A combination of `#map` and `#join` would get the result you're looking for while remaining easy to read.

File: dino_catalog/dino_dex.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
+require_relative "file_finder"
+require_relative "dino_catalog"
+require_relative "dinosaur"
+
+require "csv"
+require "json"
+
+class DinoDex
```
Comment (djkotowski) [inline dino_catalog/dino_dex.rb]: As @jmmastey, it's hard to tell just from the names what the difference is between DinoDex and DinoCatalog.

File: dino_catalog/dino_dex.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    weight_in_lbs:  :weight,
+  }
+
+  def initialize
+    @dinosaurs = []
+  end
+
+  def import_csv_file(csv_file_path)
+    extract_data(csv_file_path)
+    @dinosaurs.concat(@csv_rows.collect { |row| Dinosaur.new(row) })
```
Comment (djkotowski) [inline dino_catalog/dino_dex.rb]: Prefer `#map` over `#collect`.

See [style guide](https://github.com/bbatsov/ruby-style-guide#collections).

File: dino_catalog/dino_dex.rb [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  def select_hash(args)
+    catalog = DinoCatalog.new(dinosaurs)
+    args.each { |key, value| catalog.filter(key, value) }
+    catalog
+  end
+
+  def select_dinosaur(dinosaur_name)
+    match = dinosaurs.select { |dino| dino.name.downcase == dinosaur_name.downcase }
+    match[0]
```
Comment (djkotowski) [inline dino_catalog/dino_dex.rb]: Prefer `match.first`

File: dino_catalog/dino_dex.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def select_smaller_than(value)
+    DinoCatalog.new(@dinosaurs.select { |dinosaur| dinosaur.smaller_than?(value) })
+  end
+
+  def select_hash(args)
+    catalog = DinoCatalog.new(dinosaurs)
+    args.each { |key, value| catalog.filter(key, value) }
+    catalog
```
Comment (djkotowski) [inline dino_catalog/dino_dex.rb]: This is a good place to use `#tap`.

File: dino_catalog/dino_dex.rb [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+  def extract_data(csv_file_path)
+    @csv_rows = []
+    CSV.foreach(csv_file_path, headers: true, header_converters: :symbol) do |row|
+      @csv_rows << clean_row(row)
+    end
+  end
+
+  def clean_row(csv_row)
+    return csv_row if (HEADER_NORMALIZE.keys & csv_row.headers).empty?
```
Comment (djkotowski) [inline dino_catalog/dino_dex.rb]: It looks like you should have at least a variable combining `HEADER_NORMALIZE.keys & csv_row.headers` to DRY this up a little.

File: dino_catalog/dino_dex_shell.rb [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+  attr_accessor :working_directory, :dex, :file_finder
+
+  VALID_OPTION = /[plsaeq]/
+  PRINT_LINE_LENGTH = 40
+  PRINT_LINE_PADDING = 2
+
+  def start_shell
+    @dex = DinoDex.new
+    @file_finder = FileFinder.new(".", "*.csv")
+    @header_line_length   = 50
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: These magic numbers should be constants above too.

File: dino_catalog/dino_dex_shell.rb [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    header_line("Jurassic Park, DinoDex Interface")
+    header_line("Version 4.0.5, Alpha E")
+    header_line
+  end
+
+  def header_line(line_str = "")
+    return print_line if line_str == ""
+    empty_space = PRINT_LINE_LENGTH - line_str.length - (PRINT_LINE_PADDING * 2)
+    return puts "#" * PRINT_LINE_LENGTH if empty_space < 0
+    line_str = "##" + (" " * (empty_space / 2)) + line_str + (" " * (empty_space / 2))
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: There is a bit much going on in this line.  Try splitting it up and DRY it out.

File: dino_catalog/dino_dex_shell.rb [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    header_line
+    header_line("Jurassic Park, DinoDex Interface")
+    header_line("Version 4.0.5, Alpha E")
+    header_line
+  end
+
+  def header_line(line_str = "")
+    return print_line if line_str == ""
+    empty_space = PRINT_LINE_LENGTH - line_str.length - (PRINT_LINE_PADDING * 2)
+    return puts "#" * PRINT_LINE_LENGTH if empty_space < 0
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: This should be
```ruby
return print_line if empty_space < 0
```

File: dino_catalog/dino_dex_shell.rb [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+  end
+
+  def parse_option(input)
+    input = input.strip.downcase
+    return unless valid_option?(input)
+    input
+  end
+
+  def valid_option?(input)
+    valid = VALID_OPTION =~ input
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: To make this a boolean, use `!!` to double-negate it.  Then you can simplify the below lines to
```ruby
puts "Not a valid option." unless valid
valid
```

Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: As a side note, I'd separate the console output from the predicate method.  This method should have one responsibility.

File: dino_catalog/dino_dex_shell.rb [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+  def prompt_set_path
+    printf "Please specify the CSV search path or (!Q)uit: "
+    input = gets.strip
+    case input.downcase
+      when "!q" then print_main
+      else
+        prompt_set_path unless set_working_dir(input)
+        @file_finder.search
+        print_main
+    end
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: Replace case statement with guard clause, as there are only two possible paths.

File: dino_catalog/dino_dex_shell.rb [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+      when "!q" then print_main
+      else
+        clear_screen
+        print_header
+        puts "\nDinosaur Search Results:\n-----"
+        search_hash = parse_search_options(input)
+        @dex.select_hash(search_hash).print_data(search_hash.keys.unshift("name"))
+        print_line
+        puts "\n\n"
+    end
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: See comment on L107.

File: dino_catalog/dino_dex_shell.rb [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+  def prompt_load_csv
+    puts "Loading CSV files from working path..."
+    @file_finder.search
+    if file_finder.files.empty?
+      clear_screen
+      print_header
+      puts "No CSV files found in current directory. Try changing the path and load again.\n\n"
+    else
+      clear_screen
+      print_header
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: DRY this method out by moving these lines above the if-statement.

File: dino_catalog/file_finder.rb [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    Dir.chdir(@directory_path)
+    @directory_path = File.join(Dir.pwd, "/")
+    search
+  end
+
+  def directory_path=(path)
+    return puts ERROR_NO_DIRECTORY unless Dir.exist?(path)
+    Dir.chdir(path)
+    @directory_path = File.join(Dir.pwd, "/")
+    true
```
Comment (djkotowski) [inline dino_catalog/file_finder.rb]: Accessors should return the right side of the expression, not `true`.

File: dino_catalog/Dinosaur.rb [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    return false if weight.nil?
+    !larger_than?(size)
+  end
+
+  def export_facts(facts = "")
+    to_hash.each do |header, value|
+      next if value.empty?
+      header = header.gsub(/[A-Za-z']+/, &:capitalize)
+      facts << sprintf("%-15s %s\n", header, value)
+    end
```
Comment (djkotowski) [inline dino_catalog/Dinosaur.rb]: `#each_with_object` would be more appropriate here.

File: dino_catalog/dino_catalog.rb [hunk 22]
```diff
@@ -0,0 +1,29 @@
+class DinoCatalog
+  attr_accessor :dinosaurs
```
Comment (djkotowski) [inline dino_catalog/dino_catalog.rb]: Should this be an accessor or a reader?

File: dino_catalog/dino_dex_shell.rb [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    return printf(line_str + " ##\n") unless empty_space.even?
+    printf(line_str + "##\n")
+  end
+
+  def print_line
+    puts "#" * PRINT_LINE_LENGTH
+  end
+
+  def clear_screen
+    system("clear") || system("cls")
```
Comment (djkotowski) [inline dino_catalog/dino_dex_shell.rb]: Nice cross-platform support.

File: dino_catalog/dino_dex.rb [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+
+  def select_larger_than(value)
+    DinoCatalog.new(@dinosaurs.select { |dinosaur| dinosaur.larger_than?(value) })
+  end
+
+  def select_smaller_than(value)
+    DinoCatalog.new(@dinosaurs.select { |dinosaur| dinosaur.smaller_than?(value) })
+  end
+
+  def select_hash(args)
```
Comment (djkotowski) [inline dino_catalog/dino_dex.rb]: I would replace `#select` with this and use the splat operator for args.  Then, you'll be able to filter like so:
```ruby
dinodex.select(name: "Some dinosaur")
dinodex.select(name: "Some dinosaur", weight: 3)
```

File: dino_catalog/Dinosaur.rb [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+    facts
+  end
+
+  def to_hash
+    instance_variables.each_with_object({}) do |var, data|
```
Comment (paulghaddad) [inline dino_catalog/Dinosaur.rb]: Good use of `each_with_object`.

File: dino_catalog/dino_catalog.rb (hunk 22 above)
Comment (paulghaddad) [inline dino_catalog/dino_catalog.rb]: Provide the lowest access level that's required.

File: dino_catalog/dino_dex.rb [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def select_dinosaur(dinosaur_name)
+    match = dinosaurs.select { |dino| dino.name.downcase == dinosaur_name.downcase }
+    match[0]
+  end
+
+  def to_json
+    {
+      dinosaurs: dinosaurs.map(&:to_hash),
```
Comment (paulghaddad) [inline dino_catalog/dino_dex.rb]: Good idiom with `&:to_hash`.

File: dino_catalog/dino_dex_shell.rb (hunk 17 above)
Comment (paulghaddad) [inline dino_catalog/dino_dex_shell.rb]: I agree completely.

File: dino_catalog/dino_dex_shell.rb [hunk 27]
```diff
@@ -0,0 +1,205 @@
...
+    @dex.select_hash(args)
+  end
+
+  def prompt_search_dinodex
+    clear_screen
+    print_header
+    if @dex.dinosaurs.empty?
+      return puts "No dinosaurs loaded into DinoDex system... Please load from CSV files first.\n\n"
+    end
+    puts "\nSEARCH OPTIONS: name, period, diet, weight, walking"
```
Comment (paulghaddad) [inline dino_catalog/dino_dex_shell.rb]: Consider using a `HEREDOC`.

**Comment by djkotowski** [comment]
Nice refactor.  Take a look at the comments and think about how you would refactor.  No need to resubmit.
